### PR TITLE
NestedForm Form Widget

### PR DESCRIPTION
**Pull request by munxar**
**What**
This is a new form widget that will allow a form to embed another form and bind the sub forms value to a jsonable field. This can go even more levels deep.

**Why**
Imagine you want to store longitude and latitude in one jsonable field, but don't what to write custom form widget or partial. While a repeater with minItems=1, maxItems=1 would work, it stores the data as [{}], but we want {}.
Another use case would be a meta or viewBag like field, where you just want to attach data but you don't know the schema upfront (configurable, evolving) or depending on other data.

**How**
Let's say we have a table places with the fields _id_(int), _name_(text), _geo_(text). The form should have a text field for name and *two* text fields for longitude and latitude that are stored as json in the geo column.

fields.yaml
```yaml
fields:
  name:
    label: Name
    type: text
  geo:
   type: subform
   form:
      fields:
          longitude:
               label: Longitude
               type: text
         latitude:
              label: Latitude
              type: text
```

```php
class Place {
    public $table = 'locations';
    
    public $rules = [
        'name' => 'required',
        'geo.longitude' => 'required|number',
        'geo.latitude' => 'required|number',
    ];

    public $jsonable = [ 'geo'];
}
```
The model only needs to define the field as _jsonable_. Note that even validation rules can be applied to the subform!

### CODE REVIEW

**Comment by munxar** [comment]
this fields.yaml:
```yaml
fields:
    name:
        label: Name
        span: full
        type: text
    content:
        label: Subform
        type: subform
        form:
            tabs:
                fields:
                    main1:
                        label: Color
                        type: colorpicker
                        tab: Master
                    main2:
                        label: Tags
                        type: taglist
                        mode: array
            secondaryTabs:
                fields:
                    field1:
                        label: Field1
                        tab: Info
                    field2:
                        label: Field2
                        tab: Settings
            fields:
                lat:
                    label: Latitude
                    type: number
                    span: left
                lng:
                    label: Longitude
                    type: number
                    span: right
                logo:
                    label: Logo
                    type: mediafinder
                    span: left
                    mode: image
                    imageWidth: 128
```
looks like:
bildschirmfoto 2018-12-12 um 23 48 37

and the content column in the db looks like this:
```json
{
  "lat": "1.12312",
  "lng": "0.12312312",
  "logo": "\/hqdefault.jpg",
  "main1": "#8e44ad",
  "main2": [
    "one",
    "two",
    "three"
  ],
  "field1": "Test 123",
  "field2": "asd"
}
```

**Comment by LukeTowers** [comment]
@munxar what do you think about styling it as a "panel", like the recent change to the repeater styling?

**Comment by LukeTowers** [comment]
Also, I just realized that what you want is achievable by using array syntax:
```yaml
tabs:
    fields:
        content[main1]:
            label: Color
            type: colorpicker
            tab: Master
        content[main2]:
            label: Tags
            type: taglist
            mode: array
secondaryTabs:
    fields:
        content[field1]:
            label: Field1
            tab: Info
        content[field2]:
            label: Field2
            tab: Settings
fields:
    name:
        label: Name
        span: full
        type: text
    content[lat]:
        label: Latitude
        type: number
        span: left
    content[lng]:
        label: Longitude
        type: number
        span: right
    content[logo]:
        label: Logo
        type: mediafinder
        span: left
        mode: image
        imageWidth: 128
```

**Comment by munxar** [comment]
@LukeTowers absolutely, the array syntax inspired my to start this. But there is a key difference:
**reusability**. if you like to use the same schema definition on multiple places, or just want to clean up your fields.yaml you can do this in your fields.yaml using subforms:

```yaml
fields:
    name:
        label: Name
        span: full
        type: text
    content:
        label: Subform
        type: subform
        form: $/liip/test/models/item/subform.yaml
```

and the subform.yaml
```yaml
tabs:
  fields:
    main1:
      label: Color
      type: colorpicker
      tab: Master
    main2:
      label: Tags
      type: taglist
      mode: array
secondaryTabs:
    fields:
      field1:
        label: Field1
        tab: Info
      field2:
        label: Field2
        tab: Settings
fields:
    lat:
      label: Latitude
      type: number
      span: left
    lng:
      label: Longitude
      type: number
      span: right
    logo:
      label: Logo
      type: mediafinder
      span: left
      mode: image
      imageWidth: 128
```

regarding the visual representation: absolutely, the panel would make sense, because it's kind of the repeater's small sister and it would help to spot the grouping.

**Comment by munxar** [comment]
and here is the key difference with screenshot:
(*note* I added another jsonable field content2)
```yaml
fields:
    name:
        label: Name
        span: full
        type: text
    content:
        label: Left
        type: subform
        span: left
        form: $/liip/test/models/item/subform.yaml
    content2:
        label: Right
        type: subform
        span: right
        form: $/liip/test/models/item/subform.yaml
```

bildschirmfoto 2018-12-13 um 06 02 25

I know, two main tabs don't make sense here from a UX point, but you'll get the point.

**Comment by LukeTowers** [comment]
Could you do up an example with the panel styling?

**Comment by munxar** [comment]
I'm on it right now ;-)

**Comment by munxar** [comment]
@LukeTowers 
With panel style:

bildschirmfoto 2018-12-13 um 07 50 14

note: the main tabs need some more styling, but i don't have time right now. Before I investigate more into this, I'd like to know if it will be merged at all.

**Comment by munxar** [comment]
@LukeTowers Tabs style is fixed. If you like it can be merged.

bildschirmfoto 2018-12-13 um 16 54 05

**Comment by LukeTowers** [comment]
Could we rename it to NestedForm instead of SubForm? Also, we'll need a PR to the docs documenting this formwidget before it can be merged.

**Comment by munxar** [comment]
@LukeTowers rename is done. docs are on their way.

**Comment by LukeTowers** [comment]
Great! Could you also add a usage example to the [test plugin](https://github.com/octoberrain/oc-test-plugin)?

**Comment by munxar** [comment]
@LukeTowers can you point me to a good place in the test-pugin where this could make sense,
or should I just create another model?

**Comment by w20k** [comment]
@munxar, I do think adding a new model would be a better option!

**Comment by LukeTowers** [comment]
Meh, I don't really think it matters too much. Adding a new model is extra work, you could always just add it to an existing model. Just whatever makes the most sense to you

**Comment by munxar** [comment]
@LukeTowers finally: https://github.com/octoberrain/test-plugin/pull/59

**Comment by LukeTowers** [comment]
@munxar could you change that property to `usePanelStyles` instead?

**Comment by osmanzeki** [comment]
I'd like to test this change with RainLab.Translate to see how it behaves. I haven't had the time yet but I think it would be a good idea to make sure the ML still works fine.

@LukeTowers 
Do you think enabling RainLab.Translate by default in the October.Test project could help making sure new additions work with ML in the future too or would it be too troublesome?

**Comment by LukeTowers** [comment]
Potentially, but the October.Test plugin is just a plugin, not a project. We use labs.octodock.com to test things so we just need to be testing it with the RainLab.Translate plugin included.

**Comment by munxar** [comment]
@osmanzeki thanks for testing, but can we address this in a different PR.
It's like you said, a bug how the translate plugin handles jsonable fields. 
Honestly I'd like to write a MLNestedForm version first (like the repeater), because it covers more use cases I can think of.
Then I'd glad to help you integrate your bugfix for the jsonable fields into the rainlab.translate core.

File: modules/backend/formwidgets/nestedform/partials/_nestedform.htm [hunk 1]
```diff
@@ -1,3 +1,3 @@
-<div class="subform <?= $this->isStyled ? 'styled' : '' ?>">
+<div class="subform <?= $this->usePanelStyles ? 'panel-styles' : '' ?>">
```
Comment (LukeTowers) [inline modules/backend/formwidgets/nestedform/partials/_nestedform.htm]: Need to update the `.subform` class

**Comment by munxar** [comment]
@LukeTowers ready to merge.

**Comment by LukeTowers** [comment]
@munxar works fantastic, one last change needed: Indentation on LESS file should be four spaces, not two :)

**Comment by munxar** [comment]
@LukeTowers done.

**Pull request merged**